### PR TITLE
feat(cloudwatch): support anomaly detection alarms

### LINKED ISSUE #10540: [aws-cloudwatch] Missing ThresholdMetricId
When creating Cloudwatch Alarm that makes use of anomaly detection the following error is raised at deploy time:

```
ComparisonOperators for ranges require ThresholdMetricId to be set (Service: AmazonCloudWatch; Status Code: 400; Error Code: ValidationError; Request ID: xxx; Proxy: null)
```

### Reproduction Steps

```python
aws_cloudwatch.Alarm(
     self,
    "my_alarm",
    evaluation_periods=1,
    metric=sqs_queue.metric_number_of_messages_sent(),
    threshold=2,
    statistic="Average",
    comparison_operator=aws_cloudwatch.ComparisonOperator.LESS_THAN_LOWER_OR_GREATER_THAN_UPPER_THRESHOLD,
)
```

The Alarm construct works great for any of the other ComparisonOperators that do not rely on anomaly detection i.e. `GreaterThanThreshold`. 

### Environment

  - **CLI Version      :** 1.62.0
  - **Framework Version:** 1.64.0
  - **Node.js Version:** v12.14.1
  - **OS               :** MacOS Catalina
  - **Language (Version):** Python 3.7

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by brandondahler**
* Allows customers to create Alarms with anomaly detection enabled.  
* Supports both direct metrics and math expressions.

Closes #10540

### Implementation Notes
* `TreatMissingData` has to be moved from `alarm.ts` to `metric.ts` as it not longer is specific to only `Alarm`.
* Deprecated anomaly-based comparison operators from `Alarm`'s `ComparisonOperator` enum.
  * Created a new `AnomalyDetectionComparisonOperator` enum to contain the anomaly-based comparison operators.
* `AnomalyDetectionAlarm` is mostly copied from `Alarm`, main change required was to update how renderMetric works.
  * Removed horizontalAnnotation behavior as I don't believe there's a direct equivalent for anomaly detection at the moment.
  * The copying did duplicate some utility functions, wasn't sure if it was worth moving into the `private/metric-util` file or not.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 1]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
   readonly datapointsToAlarm?: number;
 }
 
+/**
+ * Properties needed to make an alarm from a metric
+ */
+export interface CreateAnomalyDetectionAlarmOptions {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:893]: This should live in `anomaly-detection-alarm.ts` and should be renamed `AnomalyDetectionAlarmOptions`

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 2]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
   readonly datapointsToAlarm?: number;
 }
 
+/**
+ * Properties needed to make an alarm from a metric
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:891]:
```suggestion
 * Options for an AnomolyDetectionAlarm
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 3]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
...
+   * Name of the alarm
+   *
+   * @default Automatically generated name
+   */
+  readonly alarmName?: string;
+
+  /**
+   * Description for the alarm
+   *
+   * @default No description
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:905]:
```suggestion
   * @default - No description
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 4]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
...
 
+/**
+ * Properties needed to make an alarm from a metric
+ */
+export interface CreateAnomalyDetectionAlarmOptions {
+
+  /**
+   * Name of the alarm
+   *
+   * @default Automatically generated name
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:898]:
```suggestion
   * @default - Automatically generated name
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 5]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
...
+   * Description for the alarm
+   *
+   * @default No description
+   */
+  readonly alarmDescription?: string;
+
+  /**
+   * Comparison to use to check if metric is breaching
+   *
+   * @default GreaterThanOrEqualToThreshold
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:912]: GreaterThanOrEqualToThreshold does not exist on AnomalyDetectionComparisonOperator

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 6]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
...
+
+  /**
+   * The number of periods over which data is compared to the specified threshold.
+   */
+  readonly evaluationPeriods: number;
+
+  /**
+   * Sets how this alarm is to handle missing data points.
+   *
+   * @default TreatMissingData.Missing
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:931]:
```suggestion
   * @default TreatMissingData.MISSING
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 7]
```diff
@@ -823,6 +887,70 @@ export interface CreateAlarmOptions {
...
+   * @default true
+   */
+  readonly actionsEnabled?: boolean;
+
+  /**
+   * The number of datapoints that must be breaching to trigger the alarm. This is used only if you are setting an "M
+   * out of N" alarm. In that case, this value is the M. For more information, see Evaluating an Alarm in the Amazon
+   * CloudWatch User Guide.
+   *
+   * @default ``evaluationPeriods``
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:947]:
```suggestion
   * @default - same as evaluationPeriods
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 8]
```diff
@@ -677,6 +695,27 @@ export class MathExpression implements IMetric {
...
+    return new AnomalyDetectionAlarm(scope, id, {
+      metric: this,
+      alarmName: props.alarmName,
+      alarmDescription: props.alarmDescription,
+      comparisonOperator: props.comparisonOperator,
+      datapointsToAlarm: props.datapointsToAlarm,
+      threshold: props.threshold,
+      evaluationPeriods: props.evaluationPeriods,
+      treatMissingData: props.treatMissingData,
+      actionsEnabled: props.actionsEnabled,
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:714]: We should be able to just send `...props` here:

```ts
return new AnomalyDetectionAlarm(scope, id, {
  metric: this,
  ...props,
});

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 9]
```diff
@@ -677,6 +695,27 @@ export class MathExpression implements IMetric {
     });
   }
 
+  /**
+   * Make a new AnomalyDetectionAlarm for this metric
+   *
+   * Combines both properties that may adjust the metric (aggregation) as well
+   * as alarm properties.
+   */
+  public createAnomalyDetectionAlarm(scope: Construct, id: string, props: CreateAnomalyDetectionAlarmOptions): AnomalyDetectionAlarm {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:704]:
```suggestion
  public createAnomalyDetectionAlarm(scope: Construct, id: string, options: CreateAnomalyDetectionAlarmOptions): AnomalyDetectionAlarm {
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts [hunk 10]
```diff
@@ -446,6 +447,23 @@ export class Metric implements IMetric {
     });
   }
 
+  /**
+   * Make a new Anomaly Detection Alarm for this metric.
+   */
+  public createAnomalyDetectionAlarm(scope: Construct, id: string, props: CreateAnomalyDetectionAlarmOptions): AnomalyDetectionAlarm {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:453]: same comments here from the other `createAnomalyDetectionAlarm

File: packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+      // Check per-instance metric
+      const metricConfig = this.metric.toMetricConfig();
+      if (metricConfig.metricStat?.dimensions?.length != 1 || metricConfig.metricStat?.dimensions![0].name != 'InstanceId') {
+        throw new Error(`EC2 alarm actions requires an EC2 Per-Instance Metric. (${JSON.stringify(metricConfig)} does not have an 'InstanceId' dimension)`);
+      }
+    }
+    return actionArn;
+  }
+
+  private renderMetric(metric: IMetric, threshold: number) {
```
Comment (kaizen3031593) [inline packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts:143]: question: how much of this is copied directly from `Alarm.renderMetric` and how much is original?

**Review by kaizen3031593 — CHANGES_REQUESTED** [review]
Hi @brandondahler! Thanks for picking this up. The README looks excellent, and I have a few preliminary comments regarding the implementation.

**Comment by brandondahler** [comment]
I'll work through those suggestions soon.  

FYSA - In general most of this code was copied from Alarm, so the naming and documentation comments are likely also applicable there.  I'm fine with making them for this new construct but that will make this construct diverge from how the normal Alarm construct is named / documented.

File: packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+   * The metric object this alarm was based on
+   */
+  public readonly metric: IMetric;
+
+  constructor(scope: Construct, id: string, props: AnomalyDetectionAlarmProps) {
+    super(scope, id, {
+      physicalName: props.alarmName,
+    });
+
+    const comparisonOperator = props.comparisonOperator || AnomalyDetectionComparisonOperator.LESS_THAN_LOWER_OR_GREATER_THAN_UPPER_THRESHOLD;
```
[the remark this answers is not in the document]
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts:78]: Done

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 1 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:893]: Done

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 2 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:891]: Done

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 10 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:453]: Done

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 9 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:704]: Done

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 8 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:714]: Fixed

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 7 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:947]: Fixed

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 6 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:931]: Fixed

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 5 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:912]: Fixed, set to 

```
* @default - AnomalyDetectionComparisonOperator.LESS_THAN_LOWER_OR_GREATER_THAN_UPPER_THRESHOLD
```

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 4 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:898]: Fixed

File: packages/@aws-cdk/aws-cloudwatch/lib/metric.ts (hunk 3 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/metric.ts:905]: Fixed

**Comment by brandondahler** [comment]
Review comments addressed, rebased on latest master branch.

File: packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts (hunk 11 above)
Comment (brandondahler) [reply packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts:207]: It originally was copied directly from `Alarm.reenderMetric`.  I then:

1. Added code to assign/generate an id for the anomaly band expression
2. Added code to resolve the target metric's id, generating it earlier than originally if it isn't assigned in the case of a MathExpression
3. Updated the returns of the callbacks to include the `thresholdMetricId` value that was assigned/generated in the output.
4. Updated the returns of the callbacks for `dispatchMetric` to include the extra metric:
   ```
            {
              expression: `ANOMALY_DETECTION_BAND(${metricId}, ${threshold})`,
              id: thresholdMetricId,
              label: 'Expected',
              returnData: true,
            },
   ```

**Comment by brandondahler** [comment]
Looks like its just under the wire on branch coverage.

**Comment by brandondahler** [comment]
Improved unit test coverage.

**Comment by brandondahler** [comment]
Cleaned up some nits around documentation that I found after self-review.

**Comment by pflorek** [comment]
How is the usage of the `AnomalyDetectionAlarm`?

```typescript
const metric = new Metric(...);
const anomalyDetector = new CfnAnomalyDetector(this, "AnomalyDetector", {
      metricName: invocationsMetric.metricName,
      namespace: invocationsMetric.namespace,
      stat: "Sum",
    });
const alarm = new AnomalyDetectionAlarm(this, "Alarm", {
    metric: metric,
    ...
});
```

An idea, does it make sense to construct `CfnAnomalyDetector` inside `AnomalyDetectionAlarm` for the given metric?

**Comment by brandondahler** [comment]
My understanding from the documentation and testing is that CfnAnomalyDetector isn’t necessary for alarm usage.

That simplifies usage down to the same as a normal alarm.
```
const alarm = new AnomalyDetectionAlarm(this, "Alarm", {
    metric: metric,
    ...
});
```

The only significant difference is that threshold represents the number of standard deviations the band should use and the comparison operator determines whether it should use the upper, lower, or both thresholds.

**Comment by brandondahler** [comment]
> An idea, does it make sense to construct `CfnAnomalyDetector` inside `AnomalyDetectionAlarm` for the given metric?

I've looked a lot closer on how the documentation says things work around anomaly detection alarms and how that interacts with `CfnAnomalyDetector` (i.e. `PutAnomalyDetector`/`DeleteAnomalyDetector`).

I believe that effectively how CloudWatch with an anomaly detection alarm works is that each time it tries to evaluate the upper and lower bound, if the anomaly detector doesn't already exist (or was deleted) it'll call `PutAnomalyDetector` with the minimal information (i.e. the appropriate metric info, no extra configuration).  Because of the auto-creation the usage of `CfnAnomalyDetector` isn't technically required.

Since the creation isn't technically required, the question becomes "Should we create/delete it because it makes things easier for customers in some manner?"  

The documentation makes reference to deleting unused anomaly detectors due to them having a recurring cost; however, the pricing page makes no mention of ongoing cost of anomaly detector resources existing (only the +2 metric cost associated with running an alarm).  I believe the documentation that references that cost is just out of date.  Due to the lack of extra cost, I don't think it is necessary to manage creating *and deleting* the anomaly detectors that are otherwise auto-created by CloudWatch.

The only other consideration I found was that explicitly creating the anomaly detector allowed customers to exclude anomalous time ranges and configure the detector for a specific timezone (for DST purposes).  The ability to configure the anomaly detector is an important feature that helps ensure that the anomaly detector doesn't train on bad data; however, I don't think configuring it within the `AnomalyDetectionAlarm` makes sense.  My reasoning is that the `AnomalyDetectionAlarm` isn't explicitly tied to any anomaly detector instance -- the anomaly detector instance is a singleton per-metric/statistic.  If we were to create/manage the anomaly detector instance within the `AnomalyDetectionAlarm`, separate alarms that both target the same metric+stat (e.g. with different standard deviation thresholds) may either fight between each other on which one's settings are the effective settings used or even worse the removal of one would silently wipe out the configuration from the other.

TL;DR - I don't think that `AnomalyDetectionAlarm` is the right place to manage `CfnAnomalyDetector` instances and instead we should separately make an `AnomalyDetector` L2 construct in a different PR (to allow customers to configure the detector as they see fit).

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Looks like our branch change to v2 caused a branch conflict. Can you please take a look at getting that resolved?

**Comment by brandondahler** [comment]
Copy, I'll get it fixed when I get some time -- hopefully in the next week or two.

**Comment by brandondahler** [comment]
Rebased onto latest main

File: packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts [hunk 13]
```diff
@@ -51,18 +51,24 @@ export enum ComparisonOperator {
   /**
    * Specified statistic is lower than or greater than the anomaly model band.
    * Used only for alarms based on anomaly detection models
+   *
+   * @deprecated Use AnomalyDetectionAlarm instead.
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts:55]: I don't see a reason to do this. Can you explain why you want to deprecate these when they're used more widely than just in this context?

File: packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts [hunk 14]
```diff
@@ -74,31 +80,6 @@ const OPERATOR_SYMBOLS: {[key: string]: string} = {
   LessThanOrEqualToThreshold: '<=',
 };
 
-/**
- * Specify how missing data points are treated during alarm evaluation
- */
-export enum TreatMissingData {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts:80]: I don't think this move makes sense. This is directly related to creating an alarm.

File: packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+   */
+  LESS_THAN_LOWER_THRESHOLD = 'LessThanLowerThreshold',
+}
+
+/**
+ * An anomaly detection alarm on a CloudWatch metric.
+ *
+ * @resource AWS::CloudWatch::Alarm
+ */
+export class AnomalyDetectionAlarm extends AlarmBase {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cloudwatch/lib/anomaly-detection-alarm.ts:113]: It looks like a lot of this is duplicating what's in the `Alarm` class. We should abstract the duplications out of both into a class that these can extend.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is just an initial passthrough, I didn't get too into the weeds at this point. Please address the inline comments and then I'll provide a more thorough review.

Also, thank you for the extensive unit testing here!

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
If/when you rebase without a code change, please use the Mergify commands instead of merging in other ways. It will keep my review active so I'll be able to more easily see when you've made updates to PR.

**Comment by TheRealAmazonKendra** [comment]
Our autoclose/autowarn action doesn't seem to be picking this PR up. It would normally warn after 21 days and then close after 28. Since this has been dormant this long, I'm going to go ahead and close this. If you would like to continue working on it, we certainly welcome that. Please open a new PR addressing the last round of feed back and we'll re-review.

**Comment by Lykr** [comment]
Hi, is there any update to this issue? Do we plan to provide user a more convenient way to create anomaly detection alarms?
